src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToLinqMethodProvider.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Utilities;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    internal sealed class CSharpConvertLinqQueryToLinqMethodProvider : AbstractConvertLinqProvider
+    {
+        protected override IAnalyzer CreateAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+            => new CSharpAnalyzer(syntaxFacts, semanticModel);
+
+        private sealed class CSharpAnalyzer : Analyzer<QueryExpressionSyntax, ExpressionSyntax>
+        {
+            public CSharpAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+                : base(syntaxFacts, semanticModel)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_linq_query_to_linq_method;
+
+            protected override ExpressionSyntax Convert(QueryExpressionSyntax source)
+            {
+                var fromClause = source.FromClause;
+                var identifier = fromClause.Identifier;
+                var expression = fromClause.Expression.WithoutTrailingTrivia();
+                if (fromClause.Type != null)
+                {
+                    var lambda = SyntaxFactory.BinaryExpression(SyntaxKind.IsExpression, SyntaxFactory.IdentifierName(identifier), SyntaxFactory.Token(SyntaxKind.IsKeyword), fromClause.Type);
+                    expression = CreateInvocationExpression(expression, identifier, lambda, "Select");
+                }
+
+                var context = new ConversionContext(identifier, expression);
+                return ProcessQueryBody(context, source.Body);
+            }
+
+            protected override bool Validate(QueryExpressionSyntax originalMethod, ExpressionSyntax convertedMethod, CancellationToken cancellationToken)
+            {
+                var speculationAnalyzer = new SpeculationAnalyzer(originalMethod, convertedMethod, _semanticModel, cancellationToken);
+                if (speculationAnalyzer.ReplacementChangesSemantics())
+                {
+                    return false;
+                }
+
+                // TODO add more checks
```

Comment:
what sort of checks?

Reply:
We will add semantic checks later. Likely they will be necessary for the inverse conversion.

---
In reply to: [167387524](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387524) [](ancestors = 167387524)